### ISSUE TITLE
Suppress reboot on default

### ISSUE THREAD

**Issue by peterjc123** [microsoft/winget-cli#229]
# Description of the new feature/enhancement

I tried to do `winget install 7zip` and then it automatically triggers a reboot on my PC. It's not a good user experience. I think you should suppress the reboot after installation unless explicitly specified.

# Proposed technical implementation details (optional)

Either design an option `--reboot` and make it off by default or just never reboot after installation.

**Comment by Lebrack** [comment]
Are you sure? I intalled 7Zip with no reboot action. Even MSI/EXE classic (manual) installation of 7Zip, don´t triggers a reboot.

**Comment by peterjc123** [comment]
Oh, maybe I have an existing 7zip installation. But anyway, we should not trigger reboot automatically.

**Comment by madcampos** [comment]
If I reboot after a install will winget be able to return to that point and continue after the reboot?

**Comment by denelon** [comment]
@peterjc123 the current implementation of the client `0.1` doesn't support a forced reboot. If an installer does, it's outside of the winget.exe client. You may have encountered a bug. I haven't been able to reproduce this in my test VM.

**Comment by blogcraft** [comment]
It happened to me while installing Visual Studio Community and Microsoft Teams.

**Comment by chrisgalvan** [comment]
I had this issue while installing Visual Studio in my new computer. It did reboot in the middle of installation.
I use a script to setup all my dev tools/apps and everything was good and dandy until it rebooted, there was not way to resume or know what programs installed or not. 

Not sure if scripting is the way to go here or if even is a supported scenario but would be great to handle the reboots so I can finish intalling then rebooting.

**Comment by mmuffins** [comment]
The same just happened to me when upgrading manictime.manictime to 4.6.19.0.

**Comment by abbfmst** [comment]
Same issue here: I installed Jabra.Direct and was surprised to have my machine rebooting immediately.

**Comment by Masamune3210** [comment]
Just had visual c++ redist automatically reboot after upgrade

**Comment by dylif** [comment]
This needs to be fixed. How can a user safely run package upgrades in the background?

**Comment by o-l-a-v** [comment]
VMWare Workstation, ```winget upgrade --id VMware.WorkstationPlayer``` does not work.

```winget install --id VMware.WorkstationPlayer``` over existing version forced a reboot.

```winget-cli``` as current point is not usefull for automating operations, as each package and scenario needs special logic.

**Comment by FolfyBlue** [comment]
This just happened again when updating LibreOffice through winget..
```
Windows Installer requires a system restart. Product Name: LibreOffice 7.2.5.2. Product Version: 7.2.5.2. Product Language: 1033. Manufacturer: The Document Foundation. Type of System Restart: 1. Reason for Restart: 1.
The Windows Installer initiated a system restart to complete or continue the configuration of 'LibreOffice 7.2.5.2'.
```
This is rather problematic, as I was running `winget upgrade --all` and there were other packages to still download/install

**Comment by slonopotamus** [comment]
I want to add some info from #54370:

For some reason, installing Stevedore causes a reboot. But Stevedore MSI is **not forcing** a reboot. It only [**schedules** a reboot](https://github.com/slonopotamus/stevedore/blob/0.3.3/wix/main.wxs#L296). If you install Stevedore MSI manually, you're presented with a "would you like to reboot" dialog. If you install Stevedore via [Chocolatey package](https://community.chocolatey.org/packages/stevedore), it returns 3010 exit code and Chocolatey warns user that reboot is scheduled. So, I am not sure what and why transforms **scheduled** reboot into a **forced** reboot when Stevedore is installed using WinGet. But I believe this is something that users are unhappy about in https://github.com/microsoft/winget-cli/issues/229 and what needs to be fixed on WinGet side.

**Comment by denelon** [comment]
@slonopotamus, thanks for the additional context. It's possible that since we're doing a "silent" or "silent with progress" the reboot isn't scheduled, it's forced. There may be another switch we can pass or another way to call the install to delay the reboot.

**Comment by damirkasper** [comment]
please forgive the dumb question but how is "/norestart" not automatically enforced on every single MSI based package? this has been a thing for decades at this point and part of the reason MSIs are so easy and predictable to work with. why wouldn't that be the default switch/flag behavior?

to echo everyone here i ran an winget upgrade all earlier and i have the zoom outlook plugin which was one of the apps needing an update and outlook was open... forced reboot a few seconds later followed.


not to complicate things but it almost seems like a way around these issues is to give us a way to pass parameters to the install strings so we can just blanket add /norestart REBOOT=ReallySuppress etc... that a lot of common installers require to install without restarting if we're expecting the manifest creators to add /norestart to their MSI based manifest files. 


again MSI is the most common and often used windows installer payload wrapper. why would you not build your new windows "next gen" app deployer tool to not account for the most common windows based installer format options? is everyone developing this product from a macbook? have you never had to exist on a corporate primarily based windows infrastructure? if i was developing a new cli based installer framework my first 2 feature requirements would be - silent and unattended installs, and no forced reboots.


i am amazed by what this product has managed to achieve so quickly and impressively, while ignoring certain very seemingly obvious use cases.

**Comment by denelon** [comment]
Not every package has a "suppress reboot" option. We've implemented defer registration as a default behavior for MSIX packages. We needed to add a key to the manifest schema to warn users about reboots, and other intrusive behaviors like restarting explorer.exe. Those are changes are rolling out with the 1.2 client now.

Some packages "insist" on rebooting to avoid being in a torn state (like Visual Studio). Several packages depend on runtimes and libraries that require a reboot to be updated. Dependency support is in progress.

With MSI packages, we would likely need to modify the output to indicate to users that the application has requested a reboot so they can be informed. We would also want settings and an argument to disable the default behavior to allow reboot to override the default behavior.

This work item is larger in scope than just adding the default suppression for MSI.

It is possible to specify which arguments one wants executed using "--override".
>Note: "--override" would also override any other arguments in the manifest. We also have a feature on the backlog to add additional arguments rather than just override them.

**Comment by WereCatf** [comment]
Personally, I'm just happy that this is being worked on and not just getting swept under the rug. I've grown pretty wary of performing updates with winget because I just can't trust it not to suddenly decide to reboot my system and cause me to lose everything I've got open; it'll probably take me a good while to unlearn this wariness.

**Comment by slonopotamus** [comment]
> Not every package has a "suppress reboot" option.

1. `/norestart` is a builtin msiexec option. It exists independently of package.
2. The fact that there exist packages that do not allow to suppress reboot is not a good excuse to avoid fixing those that have such option
3. As was said earlier, MSI package has an option to *schedule* reboot. But current WinGet logic transforms that into a *forced* reboot, taking away an option to decide *when* to reboot from user.

**Comment by denelon** [comment]
> > Not every package has a "suppress reboot" option.
> 
> 1. `/norestart` is a builtin msiexec option. It exists independently of package.

Yes, we are aware of the `/norestart' option. It wasn't identified as a top priority when building the preview release and hasn't been a top priority since then.
> 2. The fact that there exist non-MSI packages that do not allow to suppress reboot is not a good excuse to avoid fixing those that have such option

This Issue is the feature request to change the default behavior for "all" packages, not just MSI based packages. Using the term "fixing" implies that something is broken. We're honoring the default behavior for running MSI packages. We will be changing the default behavior to "suppress reboot" as the title of the work item states.
> 3. As was said earlier, MSI package has an option to _schedule_ reboot. But current WinGet logic transforms that into a _forced_ reboot, taking away an option to decide _when_ to reboot from user.

The option isn't removed. Using "--override" gives control of any arguments/switches passed to the installer.

@damirkasper's comment motivated me to share the methodology we're using to prioritize work on the [roadmap](
https://github.com/microsoft/winget-cli/discussions/2063).

**Comment by EgorEmeliyanov** [comment]
I understand that this is being worked on, just wanted to note that the reproduce-ability of this is highly system-specific and depends on the middleware components you have installed (such as "vcredist") and the applications you have open while the installation is running. I encountered this while installing Microsoft's very own PowerToys -- its MSI exited with code 3010 and winget rushed to force-restart without any warning. I'm hoping to see this feature request implemented soon. Thank you for your work!

**Comment by n2779510** [comment]
A build script isn't possible with a reboot after some installs.  Automating Windows builds was looking promising, but mid-script reboot, then failure. 🤬

**Comment by FWest98** [comment]
I just had a case where an automatic update of Office (in the background) scheduled a reboot, which I was not aware of; then I did an winget update of a completely unrelated package that does not require a reboot, and my PC still randomly rebooted...

Until there is way around this unpredictable behaviour built-in, is there some workaround? Would something like `winget upgrade --override "/norestart"` work?

**Comment by Kissaki** [comment]
@FWest98 if the installer provides that flag then yes, that should work.

**Comment by gerardog** [comment]
For gerardog/gsudo project, I created an `.msi` installer using WIX. It so simple, it just copies an `.EXE` file to `Program Files\xxx`. Turns out that if you upgrade (interactively) to a new version while the app is running, it asks something like "Do you want to force close the app, or restart later?". 
But if you use 'msiexec /quiet' or ´/passive´, then the machine restarts suddenly. 

So, the reason this issue was 'hard' to reproduce is because target files must be in use first.

`msiexec` provides `/norestart` and `/promptrestart`. I've added '/norestart' switch to gsudo (3ca689a) and it worked, plus the `reboot required` (3010) return code was handled perfectly by WinGet, with no other change except the manifest. 

![image](https://user-images.githubusercontent.com/3901474/188696567-62c2ed2e-1b3c-4341-a17d-f163a462f07f.png)

But, we shouldn't change every manifest... Winget should use this flag by default for `.msi` files.

I've submitted PR #2499 to add `/norestart` by default.

**Comment by craigktreasure** [comment]
I just hit this after installing `Microsoft.DotNet.SDK.Preview`. Totally unexpected.

**Comment by schwit61** [comment]
Winget fails in the first impressions and human factors departments. Many users will start with a 'winget upgrade --all' and get rebooted.  Many of them will never use the app again. It's like getting bit by a dog when you first meet.

And blaming the app being installed won't fix the broken trust.

**Comment by denelon** [comment]
@schwit61 is there any chance you have the logs or know which application upgrade triggered a reboot? We've got fields in the manifest that may be able to help.

Unexpected reboots are beyond "just" disruptive. We're working on ways to identify applications that force reboot on upgrade. Several enhancements in the last couple of releases have been aimed at the `winget upgrade --all` scenario.

MSIX packages and MSI packages generally handle reboot suppression well. We're dealing most often with .exe-based installers that trigger a reboot directly, or they update a dependency without passing reboot suppression to their dependencies. It's very much trial and error trying to classify which packages exhibit this behavior.

**Comment by slonopotamus** [comment]
> Exit code: 0xbc2, restarting: Yes

Surprisingly, 0xbc2 == 3010.

**Comment by denelon** [comment]
3010 is the MSI return code for "ERROR_SUCCESS_REBOOT_REQUIRED"

https://learn.microsoft.com/en-us/windows/win32/msi/error-codes

**Comment by EgorEmeliyanov** [comment]
Yes, it's just an exit code that means "Soft reboot". This does not mean that the installer initiated a reboot -- interpreting this exit code is up to the calling system -- it can be a message box or simply an stdout notification that a reboot is recommended. Winget interprets this as "reboot asap, no prompts" -- this is very arbitrary.

The opposite of this code is 1641 -- that would mean that the installer has completed successfully and is calling InitiateSystemShutdown. In that case a reboot is justified as it's outside of winget's control.

**Comment by eirinn1975** [comment]
I just experienced a reboot after an upgrade -all. The package triggering the reboot was MS Visual C++ Redistributable Package 2013

**Comment by chylex** [comment]
Uninstalling Dokany caused a reboot. I kind of expected it, but it's still a terrible user experience, especially if I wanted to uninstall multiple packages that each need to reboot.

**Comment by denelon** [comment]
We've found a couple of packages that don't honor the reboot suppression flag. We believe MSIs are handled correctly, but several .exe's still reboot even when you tell them not to, or the MSI calling into them doesn't pass the flag along. We're having to look into some deep edge cases still.

It seems the most common case is updating a dependency that is a running process that requires reboot to get properly referenced/loaded.

**Comment by xhoy** [comment]
Hi, is there been any progress in the last 10 months on this issue? We would love to silently deploy software, but this may not trigger any reboots....

**Comment by nascentt** [comment]
> @denelon Have you guys reached out to Chocolately or AptGet (R.I.P.)? I've never experienced reboots with them.

The way Chocolatey does it is by spamming `shutdown -a` constantly whilst doing actions.

**Comment by nascentt** [comment]
> Hi, is there been any progress in the last 10 months on this issue? We would love to silently deploy software, but this may not trigger any reboots....

We've been holding off deploying this for years now because of this issue.  We're not about to switch over to an app management system that sometimes reboots machines.
How this hasn't been a critical priority issue I'll never know.

**Comment by nascentt** [comment]
Is Winget abandonware?  
No activity on this in half a year. No resolution in 4 1/2 years.
Winget is unusable in this state.

**Comment by Masamune3210** [comment]
Amazingly enough, this is not, in fact, the only issue that exists for the entire project, and figuring out why the installers are misbehaving and how to work around it isn't exactly simple. 


And no, spamming 'shutdown -a' will not work. All that does is abort any existing shutdown timer. It doesn't do anything if a timer isn't used or if something uses one of the dozen other ways that exist to cause a reboot.

**Comment by nascentt** [comment]
I was just stating what chocolatey does in response to denlon.

I'm sure there's many issues remaining in Winget, sadly this is the main one blocking it from being usable in any way.

**Comment by Masamune3210** [comment]
........many people use it all the time. I use it practically every day, to the point that I don't even bother going to sites to download things anymore, I just grab it through Uniget, which is just a GUI frontend for winget

**Comment by nascentt** [comment]
> ........many people use it all the time. I use it practically every day, to the point that I don't even bother going to sites to download things anymore, I just grab it through Uniget, which is just a GUI frontend for winget

Sure it's a great random reboot generator that also installs software. I use it at home too. Not really suitable for an enterprise environment though.

**Comment by Blacklands** [comment]
Does Winget have any real chance of dealing with this issue? It would appear to me that this is basically out of scope for it.

All it really does (for the most part) is download and run installers (or copy files and add a path to `PATH` for portable software), doesn't it? Which is not meant to downplay the work that goes into Winget - I just mean that Winget's design philosophy doesn't seem to be to actually _control_ the actual software installation process. I know it calls itself a "Package Manager", but it's really more a "software fetcher", isn't it? It doesn't host any actual packages (just install manifests), it doesn't actually "package" anything, it doesn't maintain its own list of the software it has installed (it just reads from the Windows Registry, and if software doesn't write to that correctly then Winget is helpless - see for the example the still-ongoing issues with managing Discord through Winget). There's also the ongoing issue of dealing with unwanted shortcuts that software creates because it doesn't offer the right install flags to control that... and other issues of that kind. It seems to me that the Winget development team basically just treats all of these as "not our responsibility".

So I think even something like spamming `shutdown -a` is out of the scope for Winget. It doesn't want to care about what software installers are doing, it just wants to run them. I'm not that happy with it myself, when I started using Winget I was under the impression that this is an actual package manager - but I personally no longer believe that now. _shrug_

**Comment by slonopotamus** [comment]
Well, I do not fully agree. I wrote an absolutely canonical MSI package that _scheduled_ a reboot via standard MSI mechanisms. But the way winget was invoking MSI installation converted a _scheduled_ reboot into an _immediate_ reboot. Note that both MSI and winget technologies come from Microsoft, so I believe it is winget responsibility to make them work together is an adequate way (and this has happened, though it took several years).

**Comment by denelon** [comment]
**Issue summary**
- Requests that WinGet suppress installer-initiated reboots by default, with an opt-in flag to allow reboots when desired.

**Current state — the framework is implemented:**
- WinGet does not initiate a reboot by default; use `--allow-reboot` to opt in.
- Reboot-required installer exit codes are recognized via the manifest `ExpectedReturnCodes` and handled gracefully: "reboot required to finish" is reported as a success with a warning (no automatic reboot), and a resume-after-reboot flow is available when reboots are allowed.
- The MSI case where a *scheduled* reboot was converted into an *immediate* reboot has been fixed (see the related PR below).

**Remaining gap (why this stays open):**
- Some installers still reboot even when suppression is requested — most often `.exe` installers that don't honor a no-restart switch, an MSI that invokes a nested executable without passing the flag through, or updating a running dependency (e.g. VCRedist) that forces a reboot to complete.
- These are per-package issues rooted in installer behavior and manifest metadata (correct silent/no-restart switches and `ExpectedReturnCodes`), rather than something WinGet can universally force at the client level.

**Re-scoping:**
- Keeping this open to track the long tail of packages that reboot despite suppression. Specific offenders are best addressed as per-package manifest fixes (Catalog-Health) so the correct no-restart switches and return codes are captured.
- If you hit a package that reboots unexpectedly, please share the exact package identifier and version so it can be triaged against its manifest.

**Related**
- #2499

*Drafted with the help of GitHub Copilot.*

**Comment by Kissaki** [comment]
>or updating a running dependency (e.g. VCRedist) that forces a reboot to complete.

I assume this refers to when the product installer triggers them, not when winget installs them as a dependency to the product installer? 

Does winget handle the reboot stuff for dependencies it installs the same way as the product installers?

**Comment by eldair** [comment]
EA App (ElectronicArts.EADesktop) also reboots without consent on update v13.768.7.6285-15019056 and it also removes the app. There was a similar issue I previously reported and was fixed but is probably related (https://github.com/microsoft/winget-pkgs/issues/352972)

Here is event from event viewer:

<img width="1513" height="248" alt="Image" src="https://github.com/user-attachments/assets/78b2d0c4-6fb1-4e58-b472-cfe36f1c02b5" />

**Comment by eldair** [comment]
No updates on this? IT is a disaster, how can it be allowed for an app updated via cli tool without UI to execute reboot without any warning or consent?!
